Xamarin.Forms.Core/Items/ItemsView.cs

```diff
@@ -0,0 +1,134 @@
+﻿using System;
+using System.Collections;
+using System.ComponentModel;
+using System.Diagnostics;
+using System.Text.RegularExpressions;
+using Xamarin.Forms.Platform;
+
+namespace Xamarin.Forms
+{
+	// TODO hartez 2018/06/23 13:42:22 Trying this out for a nicer read in OnElementPropertyChanged, not sure if I like it yet	
+	public static class PropertyChangedEventArgsExtensions
+	{
+		public static bool Is(this PropertyChangedEventArgs args, BindableProperty property)
+		{
+			return args.PropertyName == property.PropertyName;
+		}
+
+		public static bool IsOneOf(this PropertyChangedEventArgs args, params BindableProperty[] properties)
+		{
+			for (int n = 0; n < properties.Length; n++)
+			{
+				if (args.PropertyName == properties[n].PropertyName)
+				{
+					return true;
+				}
+			}
+
+			return false;
+		}
+	}
+
+	[RenderWith(typeof(_CarouselViewRenderer))]
+	public class CarouselView : ItemsView
+	{
+		public CarouselView()
+		{
+			ItemsLayout = new ListItemsLayout(ItemsLayoutOrientation.Horizontal)
+			{
+				SnapPointsType = SnapPointsType.MandatorySingle,
+				SnapPointsAlignment = SnapPointsAlignment.Center
+			};
+		}
+	}
+
+	public class ItemsView : View
+	{
+		// TODO hartez 2018/06/24 11:37:00 Give DisplayMemberPath some thought	
+
+		public static readonly BindableProperty EmptyViewProperty =
+			BindableProperty.Create(nameof(EmptyView), typeof(object), typeof(ItemsView), null);
+
+		public object EmptyView
+		{
+			get => GetValue(EmptyViewProperty);
+			set => SetValue(EmptyViewProperty, value);
+		}
+
+		public static readonly BindableProperty EmptyViewTemplateProperty =
+			BindableProperty.Create(nameof(EmptyView), typeof(DataTemplate), typeof(ItemsView), null);
+
+		public DataTemplate EmptyViewTemplate
+		{
+			get => (DataTemplate)GetValue(EmptyViewTemplateProperty);
+			set => SetValue(EmptyViewTemplateProperty, value);
+		}
+
+		public static readonly BindableProperty ItemsSourceProperty =
+			BindableProperty.Create(nameof(ItemsSource), typeof(IEnumerable), typeof(ItemsView), null);
+
+		public IEnumerable ItemsSource 
+		{
+			get => (IEnumerable)GetValue(ItemsSourceProperty);
+			set => SetValue(ItemsSourceProperty, value);
+		}
+
+		// TODO hartez 2018/08/29 17:35:10 Should ItemsView be abstract? With ItemsLayout as an interface?
+		// Trying to come up with a reasonable way to restrict CarouselView to ListItemsLayout(LinearLayout) 
+		// ((because setting Carousel to grid is ... weird? And by default it just won't do anything.))
+		// And allow CollectionView to use a broader set of Layout options
+		// So the Bindable property only exists at the CarouselView/CollectionView (i.e., concrete class) level
+		// but some version of IItemsLayout is still here?
```

Comment:
This could also be enforced through exceptions to restrict usage, but I'd use it as a last resort.